File: android/src/main/java/com/audiocontext/context/AudioContext.kt
package com.audiocontext.context

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.os.Build
import android.os.SystemClock
import androidx.annotation.RequiresApi
import com.audiocontext.nodes.AudioBufferSourceNode
import com.audiocontext.nodes.AudioDestinationNode
import com.audiocontext.nodes.AudioScheduledSourceNode
import com.audiocontext.nodes.GainNode
import com.audiocontext.nodes.StereoPannerNode
import com.audiocontext.nodes.filter.BiquadFilterNode
import com.audiocontext.nodes.oscillator.OscillatorNode
import com.audiocontext.utils.AudioBuffer
import com.audiocontext.utils.Constants
import com.facebook.jni.HybridData
import java.util.LinkedList

class AudioContext : BaseAudioContext {
  override val sampleRate: Int = Constants.SAMPLE_RATE
    get() = field
  override val destination: AudioDestinationNode = AudioDestinationNode(this)
    get() = field
  override var state = ContextState.RUNNING

  private val contextStartTime: Long
  private val mHybridData: HybridData?
  private val sources: MutableList<AudioScheduledSourceNode> = mutableListOf()
  private val audioTracksList = LinkedList<AudioTrack>()

  companion object {
    init {
      System.loadLibrary("react-native-audio-context")
    }
  }

  init {
    mHybridData = initHybrid()
    contextStartTime = SystemClock.elapsedRealtimeNanos()
  }

  external fun initHybrid(): HybridData?

  external fun install(jsContext: Long)

  private fun initAudioTrack(bufferSize: Int): AudioTrack {
    val audioAttributes =
      AudioAttributes
        .Builder()
        .setUsage(AudioAttributes.USAGE_MEDIA)
        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
        .build()

    val audioFormat =
      AudioFormat
        .Builder()
        .setSampleRate(this.sampleRate)
        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
        .setChannelMask(AudioFormat.CHANNEL_OUT_STEREO)
        .build()

    val audioTrack = AudioTrack(audioAttributes, audioFormat, bufferSize, AudioTrack.MODE_STREAM, AudioManager.AUDIO_SESSION_ID_GENERATE)

    if (audioTrack.state != AudioTrack.STATE_INITIALIZED) {
      throw IllegalStateException("Failed to initialize AudioTrack")
    }

    return audioTrack
  }

  override fun getCurrentTime(): Double = (SystemClock.elapsedRealtimeNanos() - contextStartTime) / 1e9

  fun getState(): String = ContextState.toString(state)

  override fun close() {
    state = ContextState.CLOSED
    sources.forEach { it.cleanup() }
    sources.clear()
    audioTracksList.forEach { it.release() }
    audioTracksList.clear()
  }

  @RequiresApi(Build.VERSION_CODES.N)
  override fun getAudioTrack(bufferSize: Int): AudioTrack {
    synchronized(audioTracksList) {
      if (audioTracksList.isNotEmpty()) {
        return audioTracksList.removeFirst()
      } else {
        val audioTrack = initAudioTrack(bufferSize)

        return audioTrack
      }
    }
  }

  override fun addAudioTrack(audioTrack: AudioTrack) {
    synchronized(audioTracksList) {
      audioTracksList.add(audioTrack)
    }
  }

  override fun createOscillator(): OscillatorNode {
    val oscillatorNode = OscillatorNode(this)
    sources.add(oscillatorNode)
    return oscillatorNode
  }

  override fun createGain(): GainNode = GainNode(this)

  override fun createStereoPanner(): StereoPannerNode = StereoPannerNode(this)

  override fun createBiquadFilter(): BiquadFilterNode = BiquadFilterNode(this)

  override fun createBufferSource(): AudioBufferSourceNode {
    val bufferSourceNode = AudioBufferSourceNode(this)
    sources.add(bufferSourceNode)
    return bufferSourceNode
  }

  override fun createBuffer(
    numOfChannels: Int,
    length: Int,
    sampleRate: Int,
  ): AudioBuffer {
    val buffer = AudioBuffer(numOfChannels, length, sampleRate)
    return buffer
  }
}

File: ios/context/AudioContext.m
#import <AudioContext.h>
#import "Constants.h"

@implementation AudioContext

- (instancetype)init
{
  if (self = [super init]) {
    self.audioEngine = [[AVAudioEngine alloc] init];
    self.audioEngine.mainMixerNode.outputVolume = 1;
    self.destination = [[AudioDestinationNode alloc] initWithContext:self];

    _contextStartTime = mach_absolute_time();
    _state = ContextStateRunning;
    _sampleRate = [Constants sampleRate];
  }

  return self;
}

- (void)close
{
  [self cleanup];
  _state = ContextStateClosed;
}

- (void)cleanup
{
  if (self.audioEngine.isRunning) {
    [self.audioEngine stop];
  }

  self.audioEngine = nil;
}

- (OscillatorNode *)createOscillator
{
  return [[OscillatorNode alloc] initWithContext:self];
}

- (GainNode *)createGain
{
  return [[GainNode alloc] initWithContext:self];
}

- (StereoPannerNode *)createStereoPanner
{
  return [[StereoPannerNode alloc] initWithContext:self];
}

- (BiquadFilterNode *)createBiquadFilter
{
  return [[BiquadFilterNode alloc] initWithContext:self];
}

- (AudioBufferSourceNode *)createBufferSource
{
  return [[AudioBufferSourceNode alloc] initWithContext:self];
}

- (RNAA_AudioBuffer *)createBuffer:(int)sampleRate length:(int)length numberOfChannels:(int)numberOfChannels
{
  return [[RNAA_AudioBuffer alloc] initWithSampleRate:sampleRate length:length numberOfChannels:numberOfChannels];
}

- (double)getCurrentTime
{
  uint64_t currentTime = mach_absolute_time();
  mach_timebase_info_data_t info;
  mach_timebase_info(&info);

  uint64_t elapsedNano = (currentTime - _contextStartTime) * info.numer / info.denom;
  return elapsedNano / 1e9;
}

- (NSString *)getState
{
  return [ContextState toString:_state];
}

- (double)getSampleRate
{
  return _sampleRate;
}

@end
